// ethereum-waffle/dist/cjs/compiler/compiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const getCompileFunction_1 = require("./getCompileFunction");
const findInputs_1 = require("./findInputs");
const findImports_1 = require("./findImports");
const loadConfig_1 = require("../config/loadConfig");
const saveOutput_1 = require("./saveOutput");
const imports_fs_1 = require("@resolver-engine/imports-fs");
const imports_1 = require("@resolver-engine/imports");
async function compileProject(configPath) {
    await compileAndSave(await loadConfig_1.loadConfig(configPath));
}
exports.compileProject = compileProject;
async function compileAndSave(config) {
    const output = await compile(config);
    await processOutput(output, config);
}
exports.compileAndSave = compileAndSave;
async function compile(config) {
    // Added support for backwards compatibillity - renamable node_modules path
    const resolver = imports_fs_1.ImportsFsEngine().addResolver(imports_fs_1.resolvers.BacktrackFsResolver(config.npmPath));
    const sources = await imports_1.gatherSources(findInputs_1.findInputs(config.sourcesPath), '.', resolver);
    return getCompileFunction_1.getCompileFunction(config)(sources, findImports_1.findImports(sources));
}
exports.compile = compile;
async function processOutput(output, config) {
    if (output.errors) {
        console.error(formatErrors(output.errors));
    }
    if (anyNonWarningErrors(output.errors)) {
        throw new Error('Compilation failed');
    }
    else {
        await saveOutput_1.saveOutput(output, config);
    }
}
function anyNonWarningErrors(errors) {
    return errors && !errors.every(utils_1.isWarningMessage);
}
function formatErrors(errors) {
    return errors.map(toFormattedMessage).join('\n');
}
function toFormattedMessage(error) {
    return typeof error === 'string' ? error : error.formattedMessage;
}
